### ISSUE TITLE
Automated test failing on macOS

### ISSUE THREAD

**Issue by smcv** [libsdl-org/SDL_image#248]
When @madebr's branch added coverage on macOS, the automated test that I recently added is failing.

**Comment by smcv** [comment]
@madebr, if you don't have a Mac where you can run this locally, you could set up the "Upload artifacts" workflow step to run on macOS as well as Linux. You'll get a zip file with the comparison images. Ideally the zip file should have a different name for each test run to avoid race conditions, which would probably mean adding a unique `artifacts_name` key to `matrix.platform` or something.

**Comment by smcv** [comment]
AVIF loading is not quite the same as the reference:

```
2022-05-20 16:35:24.987 testimage[7044:21439] ERROR: 05/20/22 16:35:24: First detected occurrence at position 0,0 with a squared RGB-difference of 77.
```

I thought the decoder was deterministic; but if it isn't, then the thing to do is to look at the results, check that they're visually indistinguishable, and if they are, increase the `Format.tolerance` to some reasonable value like I did for JPEG.

**Comment by smcv** [comment]
Loading `.cur` also doesn't match the reference:

```
2022-05-20 16:35:25.011 testimage[7044:21439] ERROR: 05/20/22 16:35:25: Comparison of pixels with allowable error of 0 failed 947 times.
2022-05-20 16:35:25.011 testimage[7044:21439] ERROR: 05/20/22 16:35:25: First detected occurrence at position 0,0 with a squared RGB-difference of 798.
```

That looks like a bigger difference - perhaps it's getting decoded incorrectly?

I suspect that decoding `.cur` is not a major use-case, so changing the `CUR` test-case to be `#if 0` or `#ifndef __darwin__` would be a reasonable workaround to get the rest of the tests running.

**Comment by smcv** [comment]
Actually, a lot of these seem to be getting mis-decoded. Are we perhaps using a special Mac-specific loader implementation?

**Comment by smcv** [comment]
```
2022-05-20 16:35:25.145 testimage[7044:21439] ERROR: 05/20/22 16:35:25: Assert 'TIF loading should be enabled': Failed
```

This one is easy: if the Mac build is not meant to have TIFF support, then we want to only `export SDL_IMAGE_TEST_REQUIRE_LOAD_TIF=1` on Linux.

**Comment by slouken** [comment]
> Actually, a lot of these seem to be getting mis-decoded. Are we perhaps using a special Mac-specific loader implementation?

Yes, many image formats on Apple platforms are loaded with ImageIO, but ICO and CUR files should only use that if BMP_USES_IMAGEIO is defined, which I don't think it is, by default.

**Comment by smcv** [comment]
This should help to determine whether the macOS build is broken, or just needs a bigger tolerance: https://github.com/smcv/SDL_image/suites/6596157923/artifacts/247601499

Corresponding log so you can see what options were used: https://github.com/smcv/SDL_image/runs/6529455984

**Comment by smcv** [comment]
The AVIF tests (0001 to 0004) look basically the same, so I think they just want more tolerance. Reference followed by 4 test results:

![CompareSurfaces0001_Reference bmp](https://user-images.githubusercontent.com/1561141/169588041-f36fea55-139d-478b-8e3e-24d4e05c97bc.png) ![CompareSurfaces0001_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169588045-ecd403c0-c81b-4c8f-a406-d11cd9acb16f.png) ![CompareSurfaces0002_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169588052-c92ec873-5cbf-4697-975d-331f5eefbdfc.png) ![CompareSurfaces0003_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169588058-2d037b59-82bb-4c6a-a74f-fb452247f20b.png) ![CompareSurfaces0004_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169588062-7826a95d-1bc6-441b-996b-218b85006943.png)

**Comment by smcv** [comment]
0009 (loading `sample.cur` with `SDL_RWFromFile()`, I think) looks like a genuine bug somewhere:

![CompareSurfaces0009_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169588234-f3ad7815-b623-42e1-82a7-29058b37b8ca.png)

Perhaps macOS ImageIO is misidentifying it?

**Comment by smcv** [comment]
0013 to 0016 (GIF with `SDL_RWFromFile()`) I can't see an obvious difference, but the log says the top left pixel is different:

![CompareSurfaces0013_Reference bmp](https://user-images.githubusercontent.com/1561141/169588993-99746349-e426-4d9d-b0c1-04a5aeb350e5.png) ![CompareSurfaces0013_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169588995-26713939-8abf-44ab-bf92-19e2840c68e0.png)

**Comment by smcv** [comment]
0017 (loading `sample.ico` with `SDL_RWFromFile()`) looks like the same failure more as for `.cur`:

![CompareSurfaces0017_Reference bmp](https://user-images.githubusercontent.com/1561141/169589288-b0071ef2-4584-4bb2-a702-115af83cca66.png) ![CompareSurfaces0017_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169589291-b36f979b-9c68-404e-b7c8-f12c02c361d2.png)

**Comment by smcv** [comment]
0025 (saving and re-loading a JPEG) looks like a genuine bug somewhere:

![CompareSurfaces0025_Reference bmp](https://user-images.githubusercontent.com/1561141/169589515-7f30c474-8fa0-4fa1-8aef-088848250ef2.png) ![CompareSurfaces0025_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169589519-f6bfa274-d013-4c66-8fb8-7beb108adb71.png)

**Comment by smcv** [comment]
0026 looks the same as 0025.

**Comment by smcv** [comment]
0027 to 0030 (loading `.jxl`) just need a bit more tolerance. Reference and 4 outputs:

![CompareSurfaces0027_Reference bmp](https://user-images.githubusercontent.com/1561141/169589774-3a80c0e6-3cac-469a-86eb-606e7b29267a.png) ![CompareSurfaces0027_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169589775-c44acca3-bd05-4331-84a0-1e4778199234.png) ![CompareSurfaces0028_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169589777-824307c4-aba2-4419-873f-486361e9c836.png) ![CompareSurfaces0029_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169589778-4d6e7e6e-244d-4d63-a261-3c1cdbb3c8c8.png) ![CompareSurfaces0030_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169589780-4815d208-1eea-413f-8a6b-c2187338f0eb.png)

**Comment by sezero** [comment]
> 0025 (saving and re-loading a JPEG) looks like a genuine bug somewhere:
> 
> ![CompareSurfaces0025_Reference bmp](https://user-images.githubusercontent.com/1561141/169589515-7f30c474-8fa0-4fa1-8aef-088848250ef2.png) ![CompareSurfaces0025_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169589519-f6bfa274-d013-4c66-8fb8-7beb108adb71.png)

If saving is done via the tiny_jpg option, https://github.com/libsdl-org/SDL_image/commit/f69cdef846c39ca485672fcae8a78fdd1a3ce3dd was a big patch by me, check with it just in case I screwed up something in there.

**Comment by smcv** [comment]
> 0013 to 0016 (GIF) I can't see an obvious difference, but the log says the top left pixel is different

I'm also seeing a difference reported in a Debian autobuilder environment, although annoyingly, not in my development environment where I can inspect the resulting images.

I think testing 8-bit paletted images might have been too ambitious, and I should probably back that part out so that we can rely on the test coverage of the bits that *do* work.

**Comment by smcv** [comment]
> 0009 (loading `sample.cur` with `IMG_Load()`, I think) looks like a genuine bug somewhere:
> 
> ![CompareSurfaces0009_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169588234-f3ad7815-b623-42e1-82a7-29058b37b8ca.png)

#257 

> 0025 (saving and re-loading a JPEG) looks like a genuine bug somewhere:
> 
> ![CompareSurfaces0025_Reference bmp](https://user-images.githubusercontent.com/1561141/169589515-7f30c474-8fa0-4fa1-8aef-088848250ef2.png) ![CompareSurfaces0025_TestOutput bmp](https://user-images.githubusercontent.com/1561141/169589519-f6bfa274-d013-4c66-8fb8-7beb108adb71.png)

#256 

For now, I'm going to work around those two so we can bring test coverage back online.

> The AVIF tests (0001 to 0004) look basically the same, so I think they just want more tolerance

I'll add tolerance.

> 0013 to 0016 (GIF) I can't see an obvious difference, but the log says the top left pixel is different

My code to convert to RGBA for comparison was broken. I'll fix this soon.

> 0027 to 0030 (loading `.jxl`) just need a bit more tolerance

I'll add tolerance.

**Comment by smcv** [comment]
I'm hoping #258 will fix all the parts of this that are not #256 or #257, and work around those two.

**Comment by smcv** [comment]
#258 fixes this. It no longer contains workarounds for #256 or #257, which are fixed already.
